raster/src/main/scala/geotrellis/raster/summary/CellVisitor.scala

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2019 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.raster.summary
+
+/** Visitor for cell values of T that should record their state of R
+  *
+  * The user should implement concrete subclasses that update the value of `result` as
+  * necessary on each call to `visit(raster: T, col: Int, row: Int)`.
+  *
+  * Be sure to handle the empty state. This could occur if no points in T are ever visited.
+  */
+trait CellVisitor[-T, R] {
```

Comment:
To refresh my memory why did we settle on contra-variance of `T`? Might be worth while calling out in the docs as it would impact its usage.

Reply:
Oof. Also stretching my memory here. Might have to discuss offline to build some notes for discussion in docstrings. I remember some trouble with `-CellGrid[Int]` but not specifically what it was. And users could theoretically construct a CellVisitor for something that isn't a Tile. It's likely a good idea to think a bit more about where this trait belongs and how generic we want it to be. In its current form it could be implemented for basically any two dimensional array T that contains values of R, not just tiles.

Reply:
Renamed `CellVisitor` to `GridVisitor` because our only constraint on T is that it contain cols and rows that can be iterated over and indexed into, which Grid provides. T is also now bounded on Grid[Int]. There are some new docstrings that attempt to explain why this is -T and +R, similar to Function1. @moradology and I figured that it made sense to include some external links here because we couldn't explain the reasoning in a more succinct way (and general theory doesn't typically have a place in these types of docstrings). Happy to update with any feedback you might have, this was a tough one to explain.